src/Processors/QueryPlan/Optimizations/distinctReadInOrder.cpp

```diff
@@ -1,12 +1,79 @@
 #include <memory>
 #include <Processors/QueryPlan/DistinctStep.h>
+#include <Processors/QueryPlan/ExpressionStep.h>
+#include <Processors/QueryPlan/FilterStep.h>
 #include <Processors/QueryPlan/ITransformingStep.h>
 #include <Processors/QueryPlan/Optimizations/Optimizations.h>
 #include <Processors/QueryPlan/ReadFromMergeTree.h>
 #include <Common/typeid_cast.h>
 
 namespace DB::QueryPlanOptimizations
 {
+/// build actions DAG from stack of steps
+static ActionsDAGPtr buildActionsForPlanPath(std::vector<ActionsDAGPtr> & dag_stack)
+{
+    if (dag_stack.empty())
+        return nullptr;
+
+    ActionsDAGPtr path_actions = dag_stack.back()->clone();
+    dag_stack.pop_back();
+    while (!dag_stack.empty())
+    {
+        ActionsDAGPtr clone = dag_stack.back()->clone();
+        dag_stack.pop_back();
+        path_actions->mergeInplace(std::move(*clone));
+    }
+    return path_actions;
+}
+
+static const ActionsDAG::Node * getOriginalNodeForOutputAlias(const ActionsDAGPtr & actions, const String & output_name)
+{
+    /// find alias in output
+    const ActionsDAG::Node * output_alias = nullptr;
+    for (const auto * node : actions->getOutputs())
```

Comment:
A tiny thing I don't like: we call `getOriginalNodeForOutputAlias` in a loop and do a linear search for every column in outputs.
We can build an index : output_name -> Node *
it's not very important, I think, but also easy to avoid

Reply:
Will try it in next related PR (presumably #47319)